client/allocrunner/taskrunner/task_runner_test.go

```diff
@@ -555,16 +551,188 @@ func TestTaskRunner_Dispatch_Payload(t *testing.T) {
 	require.Equal(t, expected, data)
 }
 
-// testWaitForTaskToStart waits for the task to or fails the test
-func testWaitForTaskToStart(t *testing.T, tr *TaskRunner) {
-	// Wait for the task to start
+// TestTaskRunner_SignalFailure asserts that signal errors are properly
+// propagated from the driver to TaskRunner.
+func TestTaskRunner_SignalFailure(t *testing.T) {
+	t.Parallel()
+
+	alloc := mock.Alloc()
+	task := alloc.Job.TaskGroups[0].Tasks[0]
+	task.Driver = "mock_driver"
+	errMsg := "test forcing failure"
+	task.Config = map[string]interface{}{
+		"run_for":      "10m",
+		"signal_error": errMsg,
+	}
+
+	conf, cleanup := testTaskRunnerConfig(t, alloc, task.Name)
+	defer cleanup()
+
+	tr, err := NewTaskRunner(conf)
+	require.NoError(t, err)
+	go tr.Run()
+	defer tr.Kill(context.Background(), structs.NewTaskEvent("cleanup"))
+
+	testWaitForTaskToStart(t, tr)
+
+	err = tr.Signal(&structs.TaskEvent{}, "SIGINT")
+	require.NotNil(t, err)
+	require.Equal(t, errMsg, err.Error())
```

Comment:
You can use `require.EqualError`[1] in place of these two requires.

Also, `require.Error` gives you a slightly nicer more error-specific failure message than `require.NotNil`.

[1] https://godoc.org/github.com/stretchr/testify/require#EqualError